# Network Optimization Case Study: Aiding Allies
[Source: Frederick S. Hillier and Gerald J. Lieberman.  *Introduction to Operations Research - 7th ed.*, McGraw-Hill, ISBN 0-07-232169-5]

Commander Votachev steps into the cold October night and deeply inhales the smoke from his cigarette, savoring its warmth. He surveys the destruction surrounding him — shattered windows, burning buildings, torn roads — and smiles. His two years of work training revolutionaries east of the Ural Mountains has proved successful; his troops now occupy seven strategically important cities in the Russian Federation: Kazan, Perm,Yekaterinburg, Ufa, Samara, Saratov, and Orenburg. His siege is not yet over, however. He looks to the west. Given the political and economic confusion in the Russian Federation at this time, he knows that his troops will be able to conquer Saint Petersburg and Moscow shortly. Commander Votachev will then be able to rule with the wisdom and control exhibited by his communist predecessors Lenin and Stalin.

Across the Pacific Ocean, a meeting of the top security and foreign policy advisers of the United States is in progress at the White House. The President has recently been briefed about the communist revolution masterminded by Commander Votachev and is determining a plan of action. The President reflects upon a similar October long ago in 1917, and he fears the possibility of a new age of radical Communist rule accompanied by chaos, bloodshed, escalating tensions, and possibly nuclear war. He therefore decides that the United States needs to respond and to respond quickly. Moscow has requested assistance from the United States military, and the President plans to send troops and supplies immediately.

The President turns to General Lankletter and asks him to describe the preparations being taken in the United States to send the necessary troops and supplies to the Russian Federation. General Lankletter informs the President that along with troops, weapons, ammunition, fuel, and supplies, aircraft, ships, and vehicles are being assembled at two port cities with airfields: Boston and Jacksonville. The aircraft and ships will transfer all troops and cargo across the Atlantic Ocean to the Eurasian continent. The general hands the President a list of the types of aircraft, ships, and vehicles being assembled along with a description of each type. The list is shown below.

|Transportation Type| Name| Capacity|Speed|
|:---|:---|:---|:---|
|Aircraft|C-141 Starlifter|150 tons|400 miles per hour|
|Ship|Transport|240 tons|35 miles per hour|
|Vehicle|Palletized Load System Truck|16,000 kilograms|60 miles per hour|

- All aircraft, ships, and vehicles are able to carry both troops and cargo.
- Once an aircraft or ship arrives in Europe, it stays there to support the armed forces.

The President then turns to Tabitha Neal, who has been negotiating with the NATO countries for the last several hours to use their ports and airfields as stops to refuel and resupply before heading to the Russian Federation. She informs the President that the following ports and airfields in the NATO countries will be made available to the United States military.

|Ports|Airfields|
|:---|:---|
|Napoli|London|
|Hamburg|Berlin|
|Rotterdam|Istanbul|


The President stands and walks to the map of the world projected on a large screen in the middle of the room. He maps the progress of troops and cargo from the United States to three strategic cities in the Russian Federation that have not yet been seized by Commander Votachev. The three cities are Saint Petersburg, Moscow, and Rostov. He explains that the troops and cargo will be used both to defend the Russian cities and to launch a counterattack against Votachev to recapture the cities he currently occupies. The President also explains that:

- All Starlifters and transports leave either Boston or Jacksonville.
- All transports that have traveled across the Atlantic must dock at one of the NATO ports to unload.
- Palletized load system trucks brought over in the transports will then carry all troops and materials unloaded from the ships at the NATO ports to the three strategic Russian cities not yet seized by Votachev - Saint Petersburg, Moscow, and Rostov.
- All Starlifters that have traveled across the Atlantic must land at one of the NATO airfields for refueling. The planes will then carry all troops and cargo from the NATO airfields to the three Russian cities.


<img src="images/aiding_allies_map.png" width="900" align="center">

The different routes that may be taken by the troops and supplies to reach the Russian Federation from the United States are shown below:

<img src="images/possible_routes_new.png" width="500" length="500" align="center">

## Scenario 1: When time is of the essence
Moscow and Washington do not know when Commander Votachev will launch his next attack. Leaders from the two countries have therefore agreed that troops should reach each of the three strategic Russian cities as quickly as possible.

The President has determined that the situation is so dire that cost is no object—as many Starlifters, transports, and trucks as are necessary will be used to transfer troops and cargo from the United States to Saint Petersburg, Moscow, and Rostov. Therefore,no limitations exist on the number of troops and amount of cargo that can be transferred between any cities.

The President has been given the following information about the length of the available routes between cities on the Atlantic leg:

|From|To|Length of route (km)|
|:---|:---|:---:|
|Boston |Berlin |7,250 |
|Boston |Hamburg |8,250 |
|Boston |Istanbul |8,300 |
|Boston |London |6,200 |
|Boston |Rotterdam |6,900 |
|Boston |Napoli |7,950 |
|Jacksonville |Berlin |9,200 |
|Jacksonville |Hamburg |9,800 |
|Jacksonville |Istanbul |10,100 |
|Jacksonville |London |7,900 |
|Jacksonville |Rotterdam |8,900 |
|Jacksonville |Napoli |9,400 |

and on the Eurasian leg:

|From|To|Length of route (km)|
|:---|:---|:---:|
|Berlin | Saint Petersburg |1,280 |
|Hamburg |Saint Petersburg |1,880 |
|Istanbul |Saint Petersburg |2,040 |
|London |Saint Petersburg |1,980 |
|Rotterdam |Saint Petersburg |2,200 |
|Napoli |Saint Petersburg |2,970 |
|Berlin |Moscow |1,600 |
|Hamburg |Moscow |2,120 |
|Istanbul |Moscow |1,700 |
|London |Moscow |2,300 |
|Rotterdam |Moscow |2,450 |
|Napoli |Moscow |2,890 |
|Berlin |Rostov |1,730 |
|Hamburg |Rostov |2,470 |
|Istanbul |Rostov |990 |
|London |Rostov |2,860 |
|Rotterdam |Rostov |2,760 |
|Napoli |Rostov |2,800 |

Given the distance and the speed of the transportation used between each pair of cities, how can the President most quickly move troops from the United States to each of the three strategic Russian cities?

### Dijkstra's shortest path approach
Since Question 1 is only concerned with the minimum time taken by the troops to reach Russia, we can model this problem as a *shortest path problem* with each edge length representing the time taken to travel from the source node to the target node. The problem can then be solved using *Dijkstra's shortest path algorithm*. We will use the helper classes and methods provided by the custom [solver](https://github.com/ayusbhar2/operations_research/tree/main/solver) module to solve the problem.

In [1]:
# my solver module
from solver.algorithms import get_shortest_path
from solver.classes import Edge, Graph
from solver.utils import get_result_summary, prettify

In [2]:
AIR_SPEED_KMPH = 400 * 1.60934  # from MPH to KMPH
WATER_SPEED_KMPH = 35 * 1.60934
LAND_SPEED_KMPH = 60 * 1.60934

In [3]:
# Create the network graph with provided edges
graph = Graph([
    Edge('Boston', 'Berlin', distance_km=7250, route_type='air'),
    Edge('Boston', 'Hamburg', distance_km=8250, route_type='sea'),
    Edge('Boston', 'Istanbul', distance_km=8300, route_type='air'),
    Edge('Boston', 'London', distance_km=6200, route_type='air'),
    Edge('Boston', 'Rotterdam', distance_km=6900, route_type='sea'),
    Edge('Boston', 'Napoli', distance_km=7950, route_type='sea'),
    Edge('Jacksonville', 'Berlin', distance_km=9200, route_type='air'),
    Edge('Jacksonville', 'Hamburg', distance_km=9800, route_type='sea'),
    Edge('Jacksonville', 'Istanbul', distance_km=10100, route_type='air'),
    Edge('Jacksonville', 'London', distance_km=7900, route_type='air'),
    Edge('Jacksonville', 'Rotterdam', distance_km=8900, route_type='sea'),
    Edge('Jacksonville', 'Napoli', distance_km=9400, route_type='sea'),
    
    Edge('Berlin', 'StPetersburg', distance_km=1280, route_type='air'),
    Edge('Hamburg', 'StPetersburg', distance_km=1880, route_type='land'),
    Edge('Istanbul', 'StPetersburg', distance_km=2040, route_type='air'),
    Edge('London', 'StPetersburg', distance_km=1980, route_type='air'),
    Edge('Rotterdam', 'StPetersburg', distance_km=2200, route_type='land'),
    Edge('Napoli', 'StPetersburg', distance_km=2970, route_type='land'),
    Edge('Berlin', 'Moscow', distance_km=1600, route_type='air'),
    Edge('Hamburg', 'Moscow', distance_km=2120, route_type='land'),
    Edge('Istanbul', 'Moscow', distance_km=1700, route_type='air'),
    Edge('London', 'Moscow', distance_km=2300, route_type='air'),
    Edge('Rotterdam', 'Moscow', distance_km=2450, route_type='land'),
    Edge('Napoli', 'Moscow', distance_km=2890, route_type='land'),
    Edge('Berlin', 'Rostov', distance_km=1730, route_type='air'),
    Edge('Hamburg', 'Rostov', distance_km=2470, route_type='land'),
    Edge('Istanbul', 'Rostov', distance_km=990, route_type='air'),
    Edge('London', 'Rostov', distance_km=2860, route_type='air'),
    Edge('Rotterdam', 'Rostov', distance_km=2760, route_type='land'),
    Edge('Napoli', 'Rostov', distance_km=2800, route_type='land'),
])

In [4]:
# Update all edges with a non-negative time-cost
for edge in graph.edges:
    if edge.route_type == 'air':
        time_hr = edge.distance_km / AIR_SPEED_KMPH
    elif edge.route_type == 'sea':
        time_hr = edge.distance_km / WATER_SPEED_KMPH
    else:
        time_hr = edge.distance_km / LAND_SPEED_KMPH
    edge.update(cost=time_hr) # edge cost is needed for shortest path algorithm

In [5]:
get_shortest_path(graph, 'Jacksonville', target_name='StPetersburg', algorithm='dijkstra')

(15.347906595250226, ['Jacksonville', 'London', 'StPetersburg'])

In [6]:
get_shortest_path(graph, 'Boston', target_name='StPetersburg', algorithm='dijkstra')

(12.707072464488547, ['Boston', 'London', 'StPetersburg'])

In [7]:
get_shortest_path(graph, 'Jacksonville', target_name='Moscow', algorithm='dijkstra')

(15.845004784570072, ['Jacksonville', 'London', 'Moscow'])

In [8]:
get_shortest_path(graph, 'Boston', target_name='Moscow', algorithm='dijkstra')

(13.204170653808394, ['Boston', 'London', 'Moscow'])

In [9]:
get_shortest_path(graph, 'Jacksonville', target_name='Rostov', algorithm='dijkstra')

(16.7149266158798, ['Jacksonville', 'London', 'Rostov'])

In [10]:
get_shortest_path(graph, 'Boston', target_name='Rostov', algorithm='dijkstra')

(13.949817937788161, ['Boston', 'Berlin', 'Rostov'])

From the above results we see that the fastest way to get troops and supplies to each of the three strategic Russian cities is:

- `Boston --> London --> Saint Petersburg (12.70 hr)`

- `Boston --> London --> Moscow (13.20 hr)`

- `Boston --> Berlin --> Rostov (13.95 hr)`

### Binary Integer Programming approach

We can also model the above problem as a binary integer program as follows.

Parameters:

$$c_{j, k} := \text{ the "cost" associated with edge } (j, k)$$

Variables:

$$x_i:= \begin{cases}
1 \text{ if node i is on the path}\\
0 \text{ otherwise}
\end{cases}$$

$$y_{j, k}:= \begin{cases}
1 \text{ if edge (j, k) is on the path}\\
0 \text{ otherwise}
\end{cases}
$$

Objective:

$$\text{minimize} \sum_{j, k} y_{j, k} c_{j, k}$$

Constraints:

$$x_O := 1 \qquad\text{(Origin constraint)}$$

$$x_T := 1  \qquad\text{(Target constraint)}$$

$$
\sum_{j} y_{j, i} \le 1 + M (1 - x_i)\ , \quad i \ne O \quad \text{(Inbound constraint 1)}\\
\sum_{j} y_{j, i} \ge 1 - M (1 - x_i)\ , \quad i \ne O \quad \text{(Inbound constraint 2)}
$$

$$
\sum_{k} y_{i, k} \le 1 + M (1 - x_i)\ , \quad i \ne T \quad \text{(Outbound constraint 1)}\\
\sum_{k} y_{i, k} \ge 1 - M(1 - x_i)\ , \quad i \ne T \quad \text{(Outbound constraint 2)}
$$

$$
x_j + x_k \ge 2 y_{j, k} \quad \forall\ j, k \quad \text{(Connectivity constraint)}
$$

$$ x_i , y_{j, k} \quad \text{binary}
$$

In order to understand how this formulation would work, let us consider the subgraph with a single source node, viz. Boston, and a single target node, viz. St. Petersburg.

<img src="images/bip_formulation.png" width="500">

Lets solve the above program with `cvxpy`

In [11]:
import cvxpy as cp
import numpy as np

# my custom solver module
from solver.classes import BinaryIntegerProblem
from solver.algorithms import branch_and_bound

In [12]:
def get_outbound_constraints(vertex_name, graph):
    s = '('
    t = '('
    for edge in graph.get_edges(source_name=vertex_name):
        s += 'y_{}_{} + '.format(edge.source.name, edge.target.name)
        t += 'y_{}_{} + '.format(edge.source.name, edge.target.name)
    s = s[:-2]
    s = s + ') <= 1 + M * (1 - x_{})'.format(vertex_name)
    
    t = t[:-2]
    t = t + ') >= 1 - M * (1 - x_{})'.format(vertex_name)
        
    return s, t

In [13]:
def get_inbound_constraints(vertex_name, graph):
    s = '('
    t = '('
    for edge in graph.get_edges(target_name=vertex_name):
        s += 'y_{}_{} + '.format(edge.source.name, edge.target.name)
        t += 'y_{}_{} + '.format(edge.source.name, edge.target.name)
    s = s[:-2]
    s = s + ') <= 1 + M * (1 - x_{})'.format(vertex_name)
    
    t = t[:-2]
    t = t + ') >= 1 - M * (1 - x_{})'.format(vertex_name)
        
    return s, t

In [14]:
def get_connectivity_constraints(edge):
    s = 'x_{s} + x_{t} >= 2 * y_{s}_{t}'.format(s=edge.source.name, t=edge.target.name)
    return s    

In [15]:
# Parameters

M = 1000000
for edge in graph.edges:
    exec('c_{s}_{t} = edge.cost'.format(s=edge.source.name, t=edge.target.name))

In [16]:
# Variables

variables = []
for edge in graph.edges:
    exec('y_{s}_{t} = cp.Variable(1, boolean=True, name="y_{s}_{t}")'.format(
        s=edge.source.name, t=edge.target.name))
    exec('variables.append(y_{s}_{t})'.format(s=edge.source.name, t=edge.target.name))

for vertex in graph.vertices:
    exec('x_{v} = cp.Variable(1, boolean=True, name="x_{v}")'.format(v=vertex.name))
    exec('variables.append(x_{v})'.format(v=vertex.name))

In [17]:
# Constraints

constraints_bip = [
    # origin and destination constraints
    x_Boston == 1,
    x_Jacksonville == 0,
    x_StPetersburg == 1,
    x_Moscow == 0,
    x_Rostov == 0,
]

# inbound and outbound constraints
for v in graph.vertices:
    if v.name != 'Boston' and v.name != 'Jacksonville':  # not a source node
        s, t = get_inbound_constraints(v.name, graph)
        constraints_bip.append(eval(s))
        constraints_bip.append(eval(t))
    if v.name != 'StPetersburg' and v.name != 'Moscow' and v.name != 'Rostov':  # not a target node
        s, t = get_outbound_constraints(v.name, graph)
        constraints_bip.append(eval(s))
        constraints_bip.append(eval(t))

# connectivity constraints
for edge in graph.edges:
    s = get_connectivity_constraints(edge)
    constraints_bip.append(eval(s))

In [18]:
# Objective

obj_str = ''
for edge in graph.edges:
    obj_str += 'c_{s}_{t} * y_{s}_{t} + '.format(s=edge.source.name, t=edge.target.name)

obj_str = obj_str[:-2]

objective_bip = cp.Minimize(eval(obj_str))

In [19]:
# Model

bip = BinaryIntegerProblem(objective_bip, constraints_bip)

In [20]:
bip.solve(verbose=True);

                                     CVXPY                                     
                                     v1.3.1                                    
(CVXPY) Jun 02 06:41:32 PM: Your problem has 41 variables, 69 constraints, and 0 parameters.
(CVXPY) Jun 02 06:41:32 PM: It is compliant with the following grammars: DCP, DQCP
(CVXPY) Jun 02 06:41:32 PM: (If you need to solve this problem multiple times, but with different data, consider using parameters.)
(CVXPY) Jun 02 06:41:32 PM: CVXPY will first compile your problem; then, it will invoke a numerical solver to obtain a solution.
-------------------------------------------------------------------------------
                                  Compilation                                  
-------------------------------------------------------------------------------
(CVXPY) Jun 02 06:41:32 PM: Compiling problem (target solver=GLPK_MI).
(CVXPY) Jun 02 06:41:32 PM: Reduction chain: Dcp2Cone -> CvxAttr2Constr -> ConeMatrixStuffin

In [21]:
summary = get_result_summary(bip)

In [22]:
summary['status']

'optimal'

In [23]:
summary['optimal_value']

12.707072464488547

In [24]:
df = prettify(summary['optimal_solution'])

In [25]:
#define function for conditional formatting
def cond_formatting(x):
    if type(x) == int or type(x) == float:
        if x > 0:
            return 'background-color: lightgray'
    return None
    
#display DataFrame with conditional formatting applied    
df.style.applymap(cond_formatting)

The above result confirms what we already knew from the previous section - the fastest path from Boston to StPetersburg is `Boston --> London --> Saint Petersburg (12.70 hr)`. We can solve similar fomulations for each source-target pair to get the desired results.

It is worth noting that most BIP solvers use some variation of the Branch and Bound algorithm which recursively solves repeated LP relaxations of the original BIP with the simplex method. Since the simplex algorithm is a polynomial time algorithm, the average total running time for the original BIP is close to polynomial in the size of the problem (i.e. number of variables). Let $N: = $number of vertices and $M:=$ number of edges in the graph. Then, from the above formulation, we note that the total number of variables is $M + N$. The average running time of the Branch and Bound algorithm is given by $O(M + N)^p$, $p > 1$. On the other hand, Dijkstra's shortest path algorithm has a running time of $O(MN)$.

However, while the BIP formulation only works for a single source-destination pair at a time, Dijkstra's algorithm can return the shortest path from a source node to *all* the other nodes in a single run. Hence, for larger problems, it may be advantageous to use Dijkstra's algortihm instead of a BIP formulation.


## Scenario 2: When money *is* an object

The President encounters only one problem with his first plan: he has to sell the military deployment to Congress. Under the War Powers Act, the President is required to consult with Congress before introducing troops into hostilities or situations where hostilities will occur. If Congress does not give authorization to the President for such use of troops, the President must withdraw troops after 60 days. Congress also has the power to decrease the 60-day time period by passing a concurrent resolution.

The President knows that Congress will not authorize significant spending for another country’s war, especially when voters have paid so much attention to decreasing the national debt. He therefore decides that he needs to find a way to get the needed troops and supplies to Saint Petersburg, Moscow, and Rostov at the minimum cost.

Each Russian city has contacted Washington to communicate the number of troops and supplies the city needs at a minimum for reinforcement. After analyzing the requests, General Lankletter has converted the requests from numbers of troops, gallons of gasoline, etc., to tons of cargo for easier planning. The requirements are listed below.

|City | Requirements|
|:---|:---|
|Saint Petersburg|320,000 tons|
|Moscow|440,000 tons|
|Rostov|240,000 tons|

Both in Boston and Jacksonville there are 500,000 tons of the necessary cargo available. When the United States decides to send a plane, ship, or truck between two cities, several costs occur—fuel costs, labor costs, maintenance costs, and appropriate port or airfield taxes and tariffs. These costs are listed below.

|From|To|Cost|
|:---|:---|:---|
|Boston|Berlin|\$50,000 per Starlifter|
|Boston|Hamburg|\$30,000 per transport|
|Boston |Istanbul|\$55,000 per Starlifter|
|Boston |London |\$45,000 per Starlifter|
|Boston |Rotterdam |\$30,000 per transport|
|Boston |Napoli |\$32,000 per transport|
|Jacksonville |Berlin |\$57,000 per Starlifter|
|Jacksonville |Hamburg |\$48,000 per transport|
|Jacksonville |Istanbul |\$61,000 per Starlifter|
|Jacksonville |London |\$49,000 per Starlifter|
|Jacksonville |Rotterdam |\$44,000 per transport|
|Jacksonville |Napoli |\$56,000 per transport|
|Berlin |Saint Petersburg |\$24,000 per Starlifter|
|Hamburg |Saint Petersburg |\$3,000 per truck|
|Istanbul |Saint Petersburg |\$28,000 per Starlifter|
|London |Saint Petersburg |\$22,000 per Starlifter|
|Rotterdam |Saint Petersburg |\$3,000 per truck|
|Napoli |Saint Petersburg |\$ 5,000 per truck|
|Berlin |Moscow |\$22,000 per Starlifter|
|Hamburg |Moscow |\$ 4,000 per truck|
|Istanbul |Moscow |\$25,000 per Starlifter|
|London |Moscow |\$19,000 per Starlifter|
|Rotterdam |Moscow |\$ 5,000 per truck|
|Napoli |Moscow |\$ 5,000 per truck|
|Berlin |Rostov |\$23,000 per Starlifter|
|Hamburg |Rostov |\$ 7,000 per truck|
|Istanbul |Rostov |\$ 2,000 per Starlifter|
|London |Rostov |\$ 4,000 per Starlifter|
|Rotterdam |Rostov |\$ 8,000 per truck|
|Napoli |Rostov |\$ 9,000 per truck|

The President faces a number of restrictions when trying to satisfy the requirements. Early winter weather in northern Russia has brought a deep freeze with much snow. Therefore, General Lankletter is opposed to sending truck convoys in the area.

- He convinces the President to supply Saint Petersburg only through the air.
- Moreover, the truck routes into Rostov are quite limited, so that from each port at most 2,500 trucks can be sent to Rostov.
- The Ukrainian government is very sensitive about American airplanes flying through their air space. It restricts the U.S. military to at most 200 flights from Berlin to Rostov and to at most 200 flights from London to Rostov. (The U.S. military does not want to fly around the Ukraine and is thus restricted by the Ukrainian limitations.)

How does the President satisfy each Russian city’s military requirements at minimum cost?

### Minimum cost flow approach

Our scenario lends itself well to a standard minimum cost flow formulation. Boston and Jacksonville are the *supply nodes*, Saint Petersburg, Moscow and Rostov are the *demand nodes* and every other node is a *transshipment node*. Each edge has a cost per unit of cargo associated with it, and some of the edges have maximum capacities as specified in the scenario.

<img src="images/possible_routes_new.png" width="500" length="500" align="center">



Let us write the mathematical model for this minimum cost flow problem.

Parameters:
$$n := \text{number of nodes in the network}$$

$$c_{i, j}:= \text{cost (\$ / ton) of transportation along (i, j)}$$

$$u_{i, j} := \text{maximum capacity (ton) of the edge (i, j)}$$

$$b_i:= \text{net amount of cargo (tons) generated at node i}$$


Variables:
$$x_{i, j} := \text{tons of cargo transported along the edge (i, j)}$$


Objective:
$$\text{minimize}\ \sum_{i=1}^n \sum_{j=1}^n c_{i, j} x_{i, j}$$

Constraints:
$$\sum_{j=1}^n x_{i,j} - \sum_{j=1}^n x_{j, i} = b_i\ \quad i = 1, ..., n \quad \text{(Node Flow)}$$
$$x_{i, j} \le u_{i, j} \quad \text{(Capacity)}$$ 

$$x_{i, j}\ge 0\quad \text{(Non-negativity)}$$

In [26]:
# cleanup
for edge in graph.edges:
    edge.delete_attr('cost')

In [27]:
# Parameters

VEHICLE_CAPACITY_TONS = {
    'Starlifter':150,
    'Transport': 240,
    'Truck': 16,
}
ROSTOV_TRUCK_NUM_LIMIT = 2500
BERLIN_TO_ROSTOV_FLIGHT_NUM_LIMIT = 200
LONDON_TO_ROSTOV_FLIGHT_NUM_LIMIT = 200

SUPPLY_AT_BOSTON = 500000
SUPPLY_AT_JACKSONVILLE = 500000
DEMAND_AT_STPETERSBURG = 320000
DEMAND_AT_MOSCOW = 440000
DEMAND_AT_ROSTOV = 240000

SUPPLY_NODES = ['Boston', 'Jacksonville']
DEMAND_NODES = ['StPetersburg', 'Moscow', 'Rostov']

EDGE_COSTS_PER_VEHICLE = {
    ('Boston', 'Berlin'): 50000,
    ('Boston', 'Hamburg'): 30000,
    ('Boston', 'Istanbul'): 55000,
    ('Boston', 'London'): 45000,
    ('Boston', 'Rotterdam'): 30000,
    ('Boston', 'Napoli'): 32000,
    ('Jacksonville', 'Berlin'): 57000,
    ('Jacksonville', 'Hamburg'): 48000,
    ('Jacksonville', 'Istanbul'): 61000,
    ('Jacksonville', 'London'): 49000,
    ('Jacksonville', 'Rotterdam'): 44000,
    ('Jacksonville', 'Napoli'): 56000,
    ('Berlin', 'StPetersburg'): 24000,
    ('Hamburg', 'StPetersburg'): 3000,
    ('Istanbul', 'StPetersburg'): 28000,
    ('London', 'StPetersburg'): 22000,
    ('Rotterdam', 'StPetersburg'): 3000,
    ('Napoli', 'StPetersburg'): 5000,
    ('Berlin', 'Moscow'): 22000,
    ('Hamburg', 'Moscow'): 4000,
    ('Istanbul', 'Moscow'): 25000,
    ('London', 'Moscow'): 19000,
    ('Rotterdam', 'Moscow'): 5000,
    ('Napoli', 'Moscow'): 5000,
    ('Berlin', 'Rostov'): 23000,
    ('Hamburg', 'Rostov'): 7000,
    ('Istanbul', 'Rostov'): 2000,
    ('London', 'Rostov'): 4000,
    ('Rotterdam', 'Rostov'): 8000,
    ('Napoli', 'Rostov'): 9000,
}

In [28]:
# computing edge costs

for k, v in EDGE_COSTS_PER_VEHICLE.items():  
    edge = graph.get_edge(k[0], k[1])
    cost_per_vehicle = v
    vehicle_capacity_tons = None

    if edge.route_type == 'land':
        vehicle_capacity_tons = VEHICLE_CAPACITY_TONS['Truck']
    elif edge.route_type == 'sea':
        vehicle_capacity_tons = VEHICLE_CAPACITY_TONS['Transport']
    else:
        vehicle_capacity_tons = VEHICLE_CAPACITY_TONS['Starlifter']

    cost_per_ton = cost_per_vehicle / vehicle_capacity_tons
    edge.update(cost_per_ton=cost_per_ton)

#     exec('c_{s}_{t} = edge.cost_per_ton'.format(
#         s=edge.source, t=edge.target))

In [45]:
# computing edge capacities

## start with infinite edge cpacities
for edge in graph.edges:
    edge.capacity_tons = np.inf
    
## Saint Petersburg is only reachable through air
edges_to_StPetersburg = graph.get_edges(target_name='StPetersburg')
for edge in edges_to_StPetersburg:
    if edge.route_type == 'land':
        edge.update(capacity_tons=0)

## Limited # of trucks to Rostov
edges_to_Rostov = graph.get_edges(target_name='Rostov')
for edge in edges_to_Rostov:
    if edge.route_type == 'land':
        capacity_tons = ROSTOV_TRUCK_NUM_LIMIT * VEHICLE_CAPACITY_TONS['Truck']
        edge.update(capacity_tons=capacity_tons)

## Limited # of flights over Ukrain
edge_Ber_Ros = graph.get_edge('Berlin', 'Rostov')
edge_Ber_Ros.capacity_tons = BERLIN_TO_ROSTOV_FLIGHT_NUM_LIMIT * VEHICLE_CAPACITY_TONS['Starlifter']

edge_Lon_Ros = graph.get_edge('London', 'Rostov')
edge_Lon_Ros.capacity_tons = LONDON_TO_ROSTOV_FLIGHT_NUM_LIMIT * VEHICLE_CAPACITY_TONS['Starlifter']

In [46]:
# Variables

variables = []
for edge in graph.edges:
    exec('x_{s}_{t} = cp.Variable(1, nonneg=True, name="x_{s}_{t}")'.format(
        s=edge.source.name, t=edge.target.name))
    exec('variables.append(x_{s}_{t})'.format(
        s=edge.source.name, t=edge.target.name))

[v.name() for v in variables]

['x_Boston_London',
 'x_Rotterdam_Moscow',
 'x_Hamburg_StPetersburg',
 'x_Istanbul_Moscow',
 'x_Boston_Hamburg',
 'x_Jacksonville_Hamburg',
 'x_Jacksonville_London',
 'x_Jacksonville_Berlin',
 'x_Boston_Napoli',
 'x_Hamburg_Rostov',
 'x_Berlin_Rostov',
 'x_Jacksonville_Rotterdam',
 'x_London_StPetersburg',
 'x_Hamburg_Moscow',
 'x_Napoli_StPetersburg',
 'x_London_Moscow',
 'x_Jacksonville_Napoli',
 'x_Napoli_Rostov',
 'x_Boston_Rotterdam',
 'x_Berlin_StPetersburg',
 'x_Rotterdam_StPetersburg',
 'x_London_Rostov',
 'x_Istanbul_StPetersburg',
 'x_Boston_Berlin',
 'x_Rotterdam_Rostov',
 'x_Berlin_Moscow',
 'x_Istanbul_Rostov',
 'x_Jacksonville_Istanbul',
 'x_Boston_Istanbul',
 'x_Napoli_Moscow']

In [31]:
# Objective

obj_str = ''
for edge in graph.edges:
    obj_str += '+ edge.cost_per_ton * x_{s}_{t}'.format(s=edge.source.name, t=edge.target.name)
    
exec('obj_min_cost_flow = cp.Minimize({})'.format(obj_str))

In [32]:
# Constraints

## flow constraints

node_flow_constraints = []
for v in graph.vertices:
    if v.name == 'Boston':  # supply node
        b = SUPPLY_AT_BOSTON
    elif v.name == 'Jacksonville':  # supply node
        b = SUPPLY_AT_JACKSONVILLE
    elif v.name == 'StPetersburg':  # demand node
        b =  -1 * DEMAND_AT_STPETERSBURG
    elif v.name == 'Moscow':  # demand node
        b = -1 * DEMAND_AT_MOSCOW
    elif v.name == 'Rostov':  # demand node
        b = -1 * DEMAND_AT_ROSTOV
    else:
        b = 0  # transshipment node
    v.add_attr(b=b)
        
    constr_out_str = ''
    constr_in_str = ''
    # outbound
    if v.name not in DEMAND_NODES:
        for edge in graph.get_edges(source_name=v.name):
            constr_out_str += 'x_{s}_{t} + '.format(
                s=v.name, t=edge.target.name)
        constr_out_str = constr_out_str[:-2] 
    # inbound
    if v.name not in SUPPLY_NODES:
        constr_in_str = ''
        for edge in graph.get_edges(target_name=v.name):
            constr_in_str += ' - x_{s}_{t}'.format(s=edge.source.name, t=v.name)

    constr_str = constr_out_str + constr_in_str + ' == {}'.format(b)
    constr = eval(constr_str)
    node_flow_constraints.append(constr)
    
## capacity constraints

capacity_constraints = []
for edge in graph.edges:
    constr = eval('x_{s}_{t} <= edge.capacity_tons'.format(
        s=edge.source.name, t=edge.target.name))
    capacity_constraints.append(constr)

In [44]:
for edge in graph.edges:
    print('{}, Capacity: {}'.format(edge, edge.capacity_tons))

Edge: Boston->London, Capacity: inf
Edge: Rotterdam->Moscow, Capacity: inf
Edge: Hamburg->StPetersburg, Capacity: 0
Edge: Istanbul->Moscow, Capacity: inf
Edge: Boston->Hamburg, Capacity: inf
Edge: Jacksonville->Hamburg, Capacity: inf
Edge: Jacksonville->London, Capacity: inf
Edge: Jacksonville->Berlin, Capacity: inf
Edge: Boston->Napoli, Capacity: inf
Edge: Hamburg->Rostov, Capacity: 40000
Edge: Berlin->Rostov, Capacity: 30000
Edge: Jacksonville->Rotterdam, Capacity: inf
Edge: London->StPetersburg, Capacity: inf
Edge: Hamburg->Moscow, Capacity: inf
Edge: Napoli->StPetersburg, Capacity: 0
Edge: London->Moscow, Capacity: inf
Edge: Jacksonville->Napoli, Capacity: inf
Edge: Napoli->Rostov, Capacity: 40000
Edge: Boston->Rotterdam, Capacity: inf
Edge: Berlin->StPetersburg, Capacity: inf
Edge: Rotterdam->StPetersburg, Capacity: 0
Edge: London->Rostov, Capacity: 30000
Edge: Istanbul->StPetersburg, Capacity: inf
Edge: Boston->Berlin, Capacity: inf
Edge: Rotterdam->Rostov, Capacity: 40000
Edge: 

Note that the above problem satisfies the **Feasible Solution Property** of minimum cost flow problems. viz. the problem has a feasible solution only if $\sum_{i = 1}^n b_i = 0$. Clearly in our case, this is true because the total supply at Boston and Jacksonville is equal to the total demand at St Petersburg, Moscow and Rostov, viz. $500,000$ tons.

It is also worth noting that we do not need to restrict $x_{i, j}$ to be integers. This follows from the **Integer Solutions Property** of minimum cost flow problems. This property implies that if $b_i$ and $u_{i, j}$ are integers, then so is every feasible solution of the problem. Hence, we can get away by formulating the above problem as a Linear Program and not an Integer Program! 

In [37]:
# Model

prob_min_cost_flow = cp.Problem(obj_min_cost_flow, node_flow_constraints + capacity_constraints)
prob_min_cost_flow.solve(solver=cp.GLPK, verbose=True)

                                     CVXPY                                     
                                     v1.3.1                                    
(CVXPY) Jun 02 06:43:21 PM: Your problem has 30 variables, 41 constraints, and 0 parameters.
(CVXPY) Jun 02 06:43:21 PM: It is compliant with the following grammars: DCP, DQCP
(CVXPY) Jun 02 06:43:21 PM: (If you need to solve this problem multiple times, but with different data, consider using parameters.)
(CVXPY) Jun 02 06:43:21 PM: CVXPY will first compile your problem; then, it will invoke a numerical solver to obtain a solution.
-------------------------------------------------------------------------------
                                  Compilation                                  
-------------------------------------------------------------------------------
(CVXPY) Jun 02 06:43:21 PM: Compiling problem (target solver=GLPK).
(CVXPY) Jun 02 06:43:21 PM: Reduction chain: Dcp2Cone -> CvxAttr2Constr -> ConeMatrixStuffing -

625000000.0

In [38]:
summary = get_result_summary(prob_min_cost_flow)

In [47]:
summary['status']

'optimal'

In [48]:
summary['optimal_value']

625000000.0

In [40]:
prettify(summary['optimal_solution'])

variable     value
0            x_Berlin_Moscow       0.0
1            x_Berlin_Rostov   30000.0
2      x_Berlin_StPetersburg       0.0
3            x_Boston_Berlin       0.0
4           x_Boston_Hamburg       0.0
5          x_Boston_Istanbul       0.0
6            x_Boston_London  500000.0
7            x_Boston_Napoli       0.0
8         x_Boston_Rotterdam       0.0
9           x_Hamburg_Moscow       0.0
10          x_Hamburg_Rostov   40000.0
11    x_Hamburg_StPetersburg       0.0
12         x_Istanbul_Moscow       0.0
13         x_Istanbul_Rostov   60000.0
14   x_Istanbul_StPetersburg       0.0
15     x_Jacksonville_Berlin   30000.0
16    x_Jacksonville_Hamburg   40000.0
17   x_Jacksonville_Istanbul   60000.0
18     x_Jacksonville_London  290000.0
19     x_Jacksonville_Napoli   40000.0
20  x_Jacksonville_Rotterdam   40000.0
21           x_London_Moscow  440000.0
22           x_London_Rostov   30000.0
23     x_London_StPetersburg  320000.0
24           x_Napoli_Moscow       0.0
25           x_Napoli_Rostov   40000.0
26     x_Napoli_StPetersburg       0.0
27        x_Rotterdam_Moscow       0.0
28        x_Rotterdam_Rostov   40000.0
29  x_Rotterdam_StPetersburg       0.0

Hence we conclude that the United States can meet the cargo requirements at three Russian cities at a minimum cost of **$ 625,000,000** while honoring all the specified restrictions. The corresponding optimal solution is shown below:
 <img src="images/min_cost_flow_optimal_soln.png" width="600" length="600" align="center">

## Scenario 3: When all requirements cannot be met - *prioritize!*

## Seervada Park Problem

In [35]:
import cvxpy as cp

## Seervada problem
# parameters
M = 1000000
xO = 1
xT = 1
cOA = 2
cOB = 5
cOC = 4
cAB = 2
cAD = 7
cBC = 1
cBE = 3
cBD = 4
cCE = 4
cED = 1
cDT = 5
cET = 7

# variables
xA = cp.Variable(1, boolean=True, name='xA')
xB = cp.Variable(1, boolean=True, name='xB')
xC = cp.Variable(1, boolean=True, name='xC')
xD = cp.Variable(1, boolean=True, name='xD')
xE = cp.Variable(1, boolean=True, name='xE')

yOA = cp.Variable(1, boolean=True, name='yOA')
yOB = cp.Variable(1, boolean=True, name='yOB')
yOC = cp.Variable(1, boolean=True, name='yOC')
yAB = cp.Variable(1, boolean=True, name='yAB')
yAD = cp.Variable(1, boolean=True, name='yAD')
yBC = cp.Variable(1, boolean=True, name='yBC')
yBE = cp.Variable(1, boolean=True, name='yBE')
yBD = cp.Variable(1, boolean=True, name='yBD')
yCE = cp.Variable(1, boolean=True, name='yCE')
yDT = cp.Variable(1, boolean=True, name='yDT')
yED = cp.Variable(1, boolean=True, name='yED')
yET = cp.Variable(1, boolean=True, name='yET')

# constraints
outbound_cons = [
    yOA + yOB + yOC <= 1 + M * (1 - xO),
    yOA + yOB + yOC >= 1 - M * (1 - xO),

    yAB + yAD <= 1 + M * (1 - xA),
    yAB + yAD >= 1 - M * (1 - xA),

    yBC + yBD + yBE <= 1 + M * (1 - xB),
    yBC + yBD + yBE >= 1 - M * (1 - xB),

    yCE <= 1 + M * (1 - xC),
    yCE >= 1 - M * (1 - xC),

    yDT <= 1 + M * (1 - xD),
    yDT >= 1 - M * (1 - xD),

    yED + yET <= 1 + M * (1 - xE),
    yED + yET >= 1 - M * (1 - xE),
]

inbound_cons = [
    yOA <= 1 + M * (1 - xA),
    yOA >= 1 - M * (1 - xA),
    
    yAB + yOB <= 1 + M * (1 - xB),
    yAB + yOB >= 1 - M * (1 - xB),
    
    yOC + yBC <= 1 + M * (1 - xC),
    yOC + yBC >= 1 - M * (1 - xC),

    yAD + yBD + yED <= 1 + M * (1 - xD),
    yAD + yBD + yED >= 1 - M * (1 - xD),

    yBE + yCE <= 1 + M * (1 - xE),
    yBE + yCE >= 1 - M * (1 - xE),

    yDT + yET <= 1 + M * (1 - xT),
    yDT + yET >= 1 - M * (1 - xT),
]

connectivity_cons = [
    xO + xA >= 2 * yOA,
    xO + xB >= 2 * yOB,
    xO + xC >= 2 * yOC,
    xA + xB >= 2 * yAB,
    xA + xD >= 2 * yAD,
    xB + xC >= 2 * yBC,
    xB + xE >= 2 * yBE,
    xB + xD >= 2 * yBD,
    xC + xE >= 2 * yCE,
    xE + xD >= 2 * yED,
    xD + xT >= 2 * yDT,
    xE + xT >= 2 * yET,
]

constraints = outbound_cons + inbound_cons + connectivity_cons
# objective

obj = cp.Minimize(
    cOA * yOA + 
    cOB * yOB + 
    cOC * yOC + 
    cAB * yAB + 
    cAD * yAD + 
    cBC * yBC + 
    cBE * yBE + 
    cBD * yBD +
    cCE * yCE +
    cDE * yED +
    cDT * yDT +
    cET * yET
)

NameError: name 'cDE' is not defined

In [ ]:
prob = cp.Problem(obj, constraints)

In [ ]:
prob.solve()

In [ ]:
from solver.utils import get_result_summary

In [ ]:
get_result_summary(prob)

# The maximum flow problem
## *As a linear program*

<img src="./images/seervada_maxflow.png" width="500" length="500" align="center">

In [ ]:
# variables
xOA = cp.Variable(1, nonneg = True, name='xOA')
xOB = cp.Variable(1, nonneg = True, name='xOB')
xOC = cp.Variable(1, nonneg = True, name='xOC')

xAB = cp.Variable(1, nonneg = True, name='xAB')
xAD = cp.Variable(1, nonneg = True, name='xAD')

xBC = cp.Variable(1, nonneg = True, name='xBC')
xBD = cp.Variable(1, nonneg = True, name='xBD')
xBE = cp.Variable(1, nonneg = True, name='xBE')

xCE = cp.Variable(1, nonneg = True, name='xCE')

xDT = cp.Variable(1, nonneg = True, name='xDT')

xED = cp.Variable(1, nonneg = True, name='xED')
xET = cp.Variable(1, nonneg = True, name='xET')

In [ ]:
# constraints
node_flow_cons = [
    xOA - xAB - xAD == 0,
    xOB + xAB - xBC - xBE - xBD == 0,
    xOC + xBC - xCE == 0,
    xAD + xBD + xED - xDT == 0,
    xCE + xBE - xED - xET == 0,
]

total_flow_cons = [-xOA - xOB - xOC + xDT + xET == 0,]

edge_capacity_cons = [
    xOA <= 5,
    xOB <= 7,
    xOC <= 4,
    xAB <= 1,
    xAD <= 3,
    xBD <= 4,
    xBE <= 5,
    xBC <= 2,
    xCE <= 4,
    xDT <= 9,
    xET <= 6,
    xED <= 1,
]

In [ ]:
obj = cp.Maximize(xDT + xET)

In [ ]:
cons = node_flow_cons + total_flow_cons + edge_capacity_cons

In [ ]:
prob = cp.Problem(obj, cons)

In [ ]:
prob.solve()